### PR TITLE
feat: array contains

### PR DESCRIPTION
**Pull request by mantasu**
### Description

Adds an `ArrayContains` transformer and paired Keras layer that check whether a scalar value is contained in an array feature, outputting `1.0` if present and `0.0` otherwise.

### Keras Layer Checklist
- [x] The new Keras layer extends [BaseLayer](../blob/main/src/kamae/tensorflow/layers/base.py)
- [x] The `_call` method has been implemented in the new layer.
- [x] The `compatible_dtypes` property is defined in the new layer.
- [x] The new layer is decorated with `@tf.keras.utils.register_keras_serializable(package=kamae.__name__)`.
- [x] The new layer takes a `name`, `input_dtype`, and `output_dtype` as arguments to the constructor and that this is passed to the super constructor.
- [x] The Keras layer is serializable. I have implemented the `get_config` method.
- [x] There are unit tests of the new layer. 
- [x] There is a specific test of layer serialisation added [here](../blob/main/tests/kamae/tensorflow/test_layer_serialisation.py).
- [x] The new layer is imported in the [__init__.py](../blob/main/src/kamae/tensorflow/layers/__init__.py) file in the `layers` directory.

### Spark Transformer/Estimator Checklist
- [x] The new Spark Transformer extends [BaseTransformer](../blob/main/src/kamae/spark/transformers/base.py).
- [x] If the new transform needs a fit method, a Spark Estimator has been implemented that extends [BaseEstimator](../blob/main/src/kamae/spark/estimators/base.py).
- [x] The instructions in the above docs page have been followed for the `__init__` and `setParams` methods.
- [x] The transformer uses one of the input/output mixin classes from [base.py](../blob/main/src/kamae/spark/params/base.py).
- [x] If the new transformer requires more parameters that would need to be serialised to the Spark ML pipeline, there is a implemented parameter class by extending the `Params` class [here](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.param.Params.html).
- [x] The `compatible_dtypes` property has been implemented to specify the input/output data types that my transformer/estimator supports.
- [x] A Keras subclassed layer is returned in the transformer's `get_tf_layer` method.
- [x] There are unit tests of the new transform. In particular, there are parity tests between the Spark and Keras implementations.
- [x] The new transformer/estimator is imported in the [__init__.py](../blob/main/src/kamae/spark/transformers/__init__.py) file in the `transformers`/`estimators` directory.

### Readme Checklist
- [x] There is a new entry (alphabetical order) in the README table describing the new layer/transformer

### CODE REVIEW

**Comment by mantasu** [comment]
Could I please be also given write access so I wouldn't need to create the fork? 🙏

File: src/kamae/keras/core/layers/array_contains.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        :param inputs: List of two tensors `(array, value)` to compute membership
+        over.
+        :returns: The tensor resulting from the membership operation, with `1.0`
+        where the value is found and `0.0` otherwise.
+        """
+        if len(inputs) != 2:
+            raise ValueError(f"Expected 2 inputs, got {len(inputs)} inputs instead.")
+
+        array, value = inputs
+        any_match = ops.any(ops.equal(array, value), axis=-1, keepdims=True)
```
Comment (georyetti) [inline src/kamae/keras/core/layers/array_contains.py]: Can we make axis a param of this layer? Then its more useful for a user that just reuses keras layers directly? We can set axis=-1 from Spark side

Comment (georyetti) [reply src/kamae/keras/core/layers/array_contains.py]: Same for keepdims too pls

File: src/kamae/spark/transformers/array_contains.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            raise TypeError(f"arrayCol '{arr_c}' element must be numeric, got {elem_t}")
+
+        if not isinstance(val_t, _NUMERIC_TYPES):
+            raise TypeError(f"valueCol '{val_c}' must be numeric, got {val_t}")
+
+        output_col = (
+            F.when(F.array_contains(F.col(arr_c), F.col(val_c)), F.lit(1.0))
+            .otherwise(F.lit(0.0))
+            .cast(DoubleType())
+        )
```
Comment (georyetti) [inline src/kamae/spark/transformers/array_contains.py]: Can we reuse the helper methods we have for array computation here? Then it would support nested arrays also in Spark.

File: src/kamae/keras/core/layers/array_contains.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        """
+        return [
+            "int8",
+            "uint8",
+            "int16",
+            "uint16",
+            "int32",
+            "uint32",
+            "int64",
+            "uint64",
```
Comment (georyetti) [inline src/kamae/keras/core/layers/array_contains.py:94]: Does this work for int values? Or just int arrays? If so can we add a test please. All tests currently are just with float values

File: src/kamae/keras/core/layers/array_contains.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        over.
+        :returns: The tensor resulting from the membership operation, with `1.0`
+        where the value is found and `0.0` otherwise.
+        """
+        if len(inputs) != 2:
+            raise ValueError(f"Expected 2 inputs, got {len(inputs)} inputs instead.")
+
+        array, value = inputs
+        any_match = ops.any(ops.equal(array, value), axis=-1, keepdims=True)
+        return ops.cast(any_match, "float32")
```
Comment (georyetti) [inline src/kamae/keras/core/layers/array_contains.py]: Is it expected that array contains returns a float32? I would expect a boolean. The user can set output_dtype="float32" to get this behaviour so I would say we should preserve booleans here

**Review by georyetti — CHANGES_REQUESTED** [review]
Some comments:
- Keras layer can contain more params to make it more flexible. We can set these as fixed from Spark side
- I think this can return booleans and the user can set output dtypes for auto casting to floats
- Can we use the Spark array helper functions to support nested arrays

File: src/kamae/keras/core/layers/array_contains.py (hunk 1 above)
Comment (mantasu) [reply src/kamae/keras/core/layers/array_contains.py]: ✅

File: src/kamae/spark/transformers/array_contains.py (hunk 2 above)
Comment (mantasu) [reply src/kamae/spark/transformers/array_contains.py]: ✅

File: src/kamae/keras/core/layers/array_contains.py (hunk 3 above)
Comment (mantasu) [reply src/kamae/keras/core/layers/array_contains.py:94]: Yeah should work for both arrays and scalars, added a test

File: src/kamae/keras/core/layers/array_contains.py (hunk 4 above)
Comment (mantasu) [reply src/kamae/keras/core/layers/array_contains.py]: ✅

File: src/kamae/spark/transformers/array_contains.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+        if not isinstance(val_t, _NUMERIC_TYPES):
+            raise TypeError(f"valueCol '{val_c}' must be numeric, got {val_t}")
+
+        # Apply array_contains at the innermost level
+        contains_func = nested_transform(
+            func=lambda x: F.array_contains(x, F.col(val_c)),
+            nest_level=nesting_level - 1,
+        )
+        output_col = contains_func(F.col(arr_c))
```
Comment (georyetti) [inline src/kamae/spark/transformers/array_contains.py]: We have a helper for this that makes it a bit simpler, can we use this? `single_input_single_output_array_transform`
Also there is no need to check for numeric types as this is done by the `BaseTransformer` using `compatible_dtypes`

File: src/kamae/spark/transformers/array_contains.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        inputCols: Optional[List[str]] = None,
+        outputCol: Optional[str] = None,
+        inputDtype: Optional[str] = None,
+        outputDtype: Optional[str] = None,
+        layerName: Optional[str] = None,
+    ) -> None:
+        """
+        Initializes an ArrayContainsTransformer transformer.
+
+        :param inputCols: Input column names, given as `[arrayCol, valueCol]`.
```
Comment (georyetti) [inline src/kamae/spark/transformers/array_contains.py:100]: Do we want to allow a single input col and a constant? So we can check if `array_contains(x, -1)` for example?

**Review by georyetti — CHANGES_REQUESTED** [review]
Two comments:
1. Pedantic one about using helpers, you have used some more internal ones, we have an easier wrapper for this. 
2. Do we want to allow a pattern where the scalar value is just a normal constant. Like we do with the Divide/Multiply transformers?

File: src/kamae/spark/transformers/array_contains.py (hunk 5 above)
Comment (mantasu) [reply src/kamae/spark/transformers/array_contains.py]: ✅

File: src/kamae/spark/transformers/array_contains.py (hunk 6 above)
Comment (mantasu) [reply src/kamae/spark/transformers/array_contains.py:100]: Fair idea, generalised it

**Review by georyetti — APPROVED** [review]
LGTM

File: src/kamae/spark/transformers/array_contains.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        :param dataset: Pyspark dataframe to transform.
+        :returns: Transformed pyspark dataframe.
+        """
+        array_col, value_col = self.get_multiple_input_cols("mathFloatConstant", 2)
+        df = dataset.select(array_col, value_col)
+
+        output_col = single_input_single_output_array_transform(
+            input_col=array_col,
+            input_col_datatype=self.get_column_datatype(df, df.columns[0]),
+            func=lambda x: F.array_contains(x, value_col),
```
Comment (Sbranikas) [inline src/kamae/spark/transformers/array_contains.py:161]: Just wanted to doulbe-check, array_contains can return NULL too, not just True/False, while for the Keras side ops.any(ops.equal(...)) always gives a boolean. if the array contains even  a single NULL, array contains will result NULL overall. Is this the desired behaviour, to propagate the NULL? If not we can maybe wrap it with an explcit check?

File: src/kamae/keras/core/layers/array_contains.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            array, value = self._force_cast_to_compatible_numeric_type(array, value)
+        elif self.value_constant is None and len(inputs) == 2:
+            # Dynamic value given
+            array, value = inputs
+        elif self.value_constant is not None:
+            raise ValueError("Expected 1 input when `value_constant` is set")
+        else:
+            raise ValueError(f"Expected 2 inputs, got {len(inputs)} inputs instead")
+
+        return ops.any(ops.equal(array, value), axis=self.axis, keepdims=self.keepdims)
```
Comment (Sbranikas) [inline src/kamae/keras/core/layers/array_contains.py:120]: Is the equality check here identical to what array_contains does on the Spark side? Just trying to understand whether there's any chance of a mismatch on exact equality.

 I see above  (line 111) the cast  only runs for the constant branch, so I wasn't sure what happens when the value comes in as a separate column with a different dtype.
